### PR TITLE
Better DH primes

### PR DESCRIPTION
**Pull request by karel-m**
see #114

### CODE REVIEW

**Comment by sjaeckel** [comment]
Always when I looked at that code I stumble over the part `#ifdef FAST_PK`...
I then remembered that the key exchange with private keys sized the same as the group in use takes forever.
Here are some rough before-after figures of the patch...

old:
```
DH- 768 make_key    took         1163691 cycles
DH- 768 encrypt_key took         2282943 cycles
DH-1024 make_key    took         2643969 cycles
DH-1024 encrypt_key took         5262858 cycles
DH-1536 make_key    took         6996373 cycles
DH-1536 encrypt_key took        13618535 cycles
DH-2048 make_key    took        15096687 cycles
DH-2048 encrypt_key took        29983803 cycles
DH-3072 make_key    took        44939087 cycles
DH-3072 encrypt_key took        88671579 cycles
DH-4096 make_key    took        98002323 cycles
DH-4096 encrypt_key took       198535119 cycles
DH-6144 make_key    took       319184797 cycles
DH-6144 encrypt_key took       641200605 cycles
DH-8192 make_key    took       733965242 cycles
DH-8192 encrypt_key took      1473450067 cycles
```
new:
```
DH- 768 make_key    took          311645 cycles
DH- 768 encrypt_key took          634360 cycles
DH-1024 make_key    took          586470 cycles
DH-1024 encrypt_key took         1045533 cycles
DH-1536 make_key    took          897783 cycles
DH-1536 encrypt_key took         1953797 cycles
DH-2048 make_key    took         1855046 cycles
DH-2048 encrypt_key took         3875457 cycles
DH-3072 make_key    took         4994607 cycles
DH-3072 encrypt_key took         9962571 cycles
DH-4096 make_key    took         7907908 cycles
DH-4096 encrypt_key took        16360227 cycles
DH-6144 make_key    took        20385462 cycles
DH-6144 encrypt_key took        40276083 cycles
DH-8192 make_key    took        36967311 cycles
DH-8192 encrypt_key took        80216481 cycles
```

**Comment by sjaeckel** [comment]
regarding the removal of tom's old primes: I somehow liked them, especially since https://www.schneier.com/blog/archives/2015/10/breaking_diffie.html ...
...but to have a real mitigation for that we must provide a way to generate your own group at runtime which is currently not possible.

**Comment by sjaeckel** [comment]
I cherry-picked a8ec58fda577c45a55b1ba0ef743421a910d367a from #198 
If you think that commit is wrong we can discuss that.
The duplicate commit will be dropped as soon as one of _this PR_ and _#198_ is merged and the other is rebased on top.

**Comment by karel-m** [comment]
I think it would be better to rename keysize like this:
```patch
- int dh_make_key(prng_state *prng, int wprng, int keysize, dh_key *key)
+ int dh_make_key(prng_state *prng, int wprng, int groupsize, dh_key *key)
```

Next, I would simplify calculating the key size like this:
```C
   switch (groupsize) {
      case 96:   keysize = 30; break;   /*  768-bit => key size 240-bit */
      case 128:  keysize = 30; break;   /* 1024-bit => key size 240-bit */
      case 192:  keysize = 30; break;   /* 1536-bit => key size 240-bit */
      case 256:  keysize = 40; break;   /* 2048-bit => key size 320-bit */
      case 384:  keysize = 52; break;   /* 3072-bit => key size 416-bit */
      case 512:  keysize = 60; break;   /* 4096-bit => key size 480-bit */
      case 768:  keysize = 67; break;   /* 6144-bit => key size 536-bit */
      case 1024: keysize = 77; break;   /* 8192-bit => key size 616-bit */
      default: return CRYPT_INVALID_KEYSIZE;
   }
```

**Comment by sjaeckel** [comment]
Fine by me

**Comment by karel-m** [comment]
While we are hacking on dh_make_key we should perhaps also fix #119

File: src/pk/dh/dh.c [hunk 1]
```diff
@@ -104,33 +104,44 @@ int dh_get_size(dh_key *key)
...
    }
 
    /* find key size */
-   for (x = 0; (keysize > sets[x].size) && (sets[x].size != 0); x++);
-#ifdef FAST_PK
-   keysize = MIN(sets[x].size, 32);
-#else
-   keysize = sets[x].size;
-#endif
+   for (x = 0; ((int)keysize > sets[x].size) && (sets[x].size != 0); x++);
```
Comment (sjaeckel) [inline src/pk/dh/dh.c]: here we have to check `groupsize > sets[x].size`

File: src/pk/dh/dh.c [hunk 2]
```diff
@@ -104,33 +104,44 @@ int dh_get_size(dh_key *key)
...
    int err;
 
    LTC_ARGCHK(key  != NULL);
+   LTC_ARGCHK(prng != NULL);
+
+   /* Table of the strength estimates from https://tools.ietf.org/html/rfc3526#section-8
+    * We use "Estimate 2" to get an appropriate private key (exponent) size.
+    */
+   switch (groupsize) {
+      case 96:   keysize = 30; break;   /*  768-bit => key size 240-bit */
```
Comment (sjaeckel) [inline src/pk/dh/dh.c]: I preferred the behavior before where we were a bit more relaxed and used _the group_ >= _the requested size_.

**Comment by sjaeckel** [comment]
@ltv511 could you please also review as you opened #119

**Comment by ltv511** [comment]
@sjaeckel As long as only safe primes are used for DH key exchange in LTC (which is currently the case AFAIK), the added checks for 1 and p-1 should be sufficient!

File: src/pk/dh/dh.c [hunk 3]
```diff
@@ -355,26 +375,31 @@ int dh_shared_secret(dh_key *private_key, dh_key *public_key,
...
 
-   if ((err = mp_read_radix(p, (char *)sets[private_key->idx].prime, 64)) != CRYPT_OK)     { goto error; }
-   if ((err = mp_exptmod(public_key->y, private_key->x, p, tmp)) != CRYPT_OK)           { goto error; }
+   if ((err = mp_read_radix(p, sets[private_key->idx].prime, 16)) != CRYPT_OK)  { goto error; }
+   if ((err = mp_sub_d(p, 1, p_minus1)) != CRYPT_OK)                            { goto error; }
+   if (mp_cmp(public_key->y, p_minus1) != LTC_MP_LT || mp_cmp_d(public_key->y, 1) != LTC_MP_GT) {
+      /* reject public key with: y <= 1 OR y >= p-1 */
+      err = CRYPT_INVALID_ARG;
+      goto error;
+   };
```
Comment (sjaeckel) [inline src/pk/dh/dh.c]: trailing ;

**Pull request merged**